core/src/main/java/com/linecorp/armeria/common/sse/DefaultServerSentEvent.java

```diff
@@ -0,0 +1,157 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common.sse;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.MoreObjects;
+
+/**
+ * A default implementation of the {@link ServerSentEvent} interface.
+ *
+ * @param <T> the type of the data
+ */
+final class DefaultServerSentEvent<T> implements ServerSentEvent<T> {
+
+    @Nullable
+    private final String id;
+    @Nullable
+    private final String event;
+    @Nullable
+    private final Duration retry;
+    @Nullable
+    private final String comment;
+    @Nullable
+    private final T data;
+
+    private final Function<T, String> dataStringifier;
+
+    @Nullable
+    private volatile String dataText;
+
+    private int hashCode;
+
+    DefaultServerSentEvent(@Nullable String id,
+                           @Nullable String event,
+                           @Nullable Duration retry,
+                           @Nullable String comment,
+                           @Nullable T data,
+                           Function<T, String> dataStringifier) {
+        this.id = id;
+        this.event = event;
+        this.retry = retry;
+        this.comment = comment;
+        this.data = data;
+        this.dataStringifier = requireNonNull(dataStringifier, "dataStringifier");
+    }
+
+    @Nullable
+    @Override
+    public String id() {
+        return id;
+    }
+
+    @Nullable
+    @Override
+    public String event() {
+        return event;
+    }
+
+    @Nullable
+    @Override
+    public Duration retry() {
+        return retry;
+    }
+
+    @Nullable
+    @Override
+    public String comment() {
+        return comment;
+    }
+
+    @Nullable
+    @Override
+    public T data() {
+        return data;
+    }
+
+    @Nullable
+    @Override
+    public String dataText() {
+        if (data == null) {
+            return null;
+        }
+
+        final String dataText = this.dataText;
+        if (dataText != null) {
+            return dataText;
+        }
+
+        final String applied = dataStringifier.apply(data);
+        this.dataText = applied;
+        return applied;
+    }
+
+    @Override
+    public int hashCode() {
+        if (hashCode == 0) {
+            int hashCode = dataStringifier.hashCode();
+            hashCode = 31 * hashCode + (id != null ? id.hashCode() : 0);
+            hashCode = 31 * hashCode + (event != null ? event.hashCode() : 0);
+            hashCode = 31 * hashCode + (retry != null ? retry.hashCode() : 0);
+            hashCode = 31 * hashCode + (comment != null ? comment.hashCode() : 0);
+            hashCode = 31 * hashCode + (data != null ? data.hashCode() : 0);
+            this.hashCode = hashCode;
```

Comment:
Perhaps we can use `Objects.hash()` in this case where most fields are nullable.